File: iOS/Code/Xamarin/bugTrap/bugTrapKit/Controllers/ViewControllers/DoneDoneLoginViewController.cs
using System;
using System.Collections.Generic;
using Foundation;
using UIKit;

namespace bugTrapKit
{
	public partial class DoneDoneLoginViewController : BtTrackerLoginViewController, IUITextFieldDelegate
	{

		public override TrackerType TrackerType {
			get { return TrackerType.DoneDone; }
		}

		public DoneDoneLoginViewController (IntPtr handle) : base(handle)
		{
		}


		public override void ViewDidAppear (bool animated)
		{
			base.ViewDidAppear(animated);

			SubdomainField.BecomeFirstResponder();
		}


		public override void PopulatePlaceholderLabels (bool showPlaceholders)
		{
			SubdomainField.Placeholder = showPlaceholders ? DataKeys.Subdomain.ToString() : string.Empty;
			UsernameField.Placeholder = showPlaceholders ? DataKeys.UserName.ToString() : string.Empty;
			PasswordField.Placeholder = showPlaceholders ? DataKeys.Password.ToString() : string.Empty;
		}


		public override void KeychainDataLoaded (Dictionary<DataKeys, string> storedData)
		{
			string subdomain = string.Empty, username = string.Empty, password = string.Empty;

			if (storedData.TryGetValue(DataKeys.Subdomain, out subdomain)) SubdomainField.Text = subdomain;
			if (storedData.TryGetValue(DataKeys.UserName, out username)) UsernameField.Text = username;
			if (storedData.TryGetValue(DataKeys.Password, out password)) PasswordField.Text = password;
		}


		[Export("textFieldShouldReturn:")]
		public bool ShouldReturn (UITextField textField)
		{
			if (textField == SubdomainField) UsernameField.BecomeFirstResponder();
			if (textField == UsernameField) PasswordField.BecomeFirstResponder();
			if (textField == PasswordField) {
				PasswordField.ResignFirstResponder();
				validateValuesAndLoginAsync();
			}
			return false;
		}


		async void validateValuesAndLoginAsync ()
		{
			if (!SubdomainField.HasText) {
				SubdomainField.BecomeFirstResponder();
				return;
			}

			if (!UsernameField.HasText) {
				UsernameField.BecomeFirstResponder();
				return;
			}

			if (!PasswordField.HasText) {
				PasswordField.BecomeFirstResponder();
				return;
			}

			ActivityIndicator.StartAnimating();
			ActivityIndicator.Hidden = false;

			var trackerData = new Dictionary<DataKeys, string> {
				{ DataKeys.Subdomain, SubdomainField.Text },
				{ DataKeys.UserName, UsernameField.Text },
				{ DataKeys.Password, PasswordField.Text }
			};

			// store the data in the keychain
			Keychain.StoreKeyValues(trackerData);

			var dataValid = TrackerService.Shared.SetCurrentTrackerType(TrackerType.DoneDone);

			if (!dataValid) {
				// Log.Error("Invalid Data", trackerData);
				Console.WriteLine("Invalid Data!");
				return;
			}

			var authenticated = await TrackerService.Shared.CurrentTracker.Authenticate(trackerData);

			if (authenticated) {
			
				await PresentingViewController.DismissViewControllerAsync(true);
			
			} else {

				var messageTop = "Unable to login to DoneDone with the information provided\n{0}";

				var message = string.Format(messageTop, "\nReview Username & Password\nthen try again");

				var errorDialog = UIAlertController.Create("Error", message, UIAlertControllerStyle.Alert);
				errorDialog.AddAction(UIAlertAction.Create("OK", UIAlertActionStyle.Cancel, null));

				await PresentViewControllerAsync(errorDialog, true);

				trackerData[DataKeys.Subdomain] = string.Empty;
				trackerData[DataKeys.UserName] = string.Empty;
				trackerData[DataKeys.Password] = string.Empty;

				// clear the data in the keychain
				Keychain.StoreKeyValues(trackerData);
			}

			ActivityIndicator.StopAnimating();
		}
	}
}

File: iOS/Code/Swift/bugTrap/bugTrapKit/Trackers/DoneDone/UI/DoneDoneLoginViewController.swift
//
//  BtSettingsLoginViewController.swift
//  bugTrap
//
//  Created by Colby L Williams on 6/30/14.
//  Copyright (c) 2014 bugTrap. All rights reserved.
//

import UIKit

class DoneDoneLoginViewController: TrackerLoginController {
	
	@IBOutlet var SubdomainField: UITextField!
	@IBOutlet var UsernameField: UITextField!
	@IBOutlet var PasswordField: UITextField!
	@IBOutlet var ActivityIndicator: UIActivityIndicatorView!
	
	
    override var trackerType : TrackerType {
        get {
            return TrackerType.DoneDone
        }
    }
	

	override func viewDidAppear(animated: Bool) {
		super.viewDidAppear(animated)
		
		SubdomainField.becomeFirstResponder()
	}

	
	override func populatePlaceholderLabels(showPlaceholders: Bool) {
		SubdomainField.placeholder = showPlaceholders ? DataKeys.Subdomain.string : ""
		UsernameField.placeholder  = showPlaceholders ? DataKeys.UserName.string  : ""
		PasswordField.placeholder  = showPlaceholders ? DataKeys.Password.string  : ""
	}
	
	
    override func keychainDataLoaded(storedData: Dictionary<DataKeys, String>) {
		
		if let subdomain = storedData[DataKeys.Subdomain] {
			SubdomainField.text = subdomain
		}
		
		if let username = storedData[DataKeys.UserName] {
			UsernameField.text = username
		}
		
		if let password = storedData[DataKeys.Password] {
			PasswordField.text = password
		}
    }
	
	
	
	func textFieldShouldReturn(textField: UITextField!) -> Bool
	{
		if textField == SubdomainField {
			UsernameField.becomeFirstResponder()
		}
		
		if textField == UsernameField {
			PasswordField.becomeFirstResponder()
		}
		
		if textField == PasswordField {
			PasswordField.resignFirstResponder()
			validateValuesAndLogin()
		}
        
		return false
	}
	
	
	
	func validateValuesAndLogin() {
		
		if !SubdomainField.hasText() {
			SubdomainField.becomeFirstResponder()
			return
		}
		if !UsernameField.hasText() {
			UsernameField.becomeFirstResponder()
			return
		}
		if !PasswordField.hasText() {
			PasswordField.becomeFirstResponder()
			return
		}
		
		ActivityIndicator.startAnimating()
		ActivityIndicator.hidden = false
		
        let trackerData = [DataKeys.Subdomain: SubdomainField.text!, DataKeys.UserName: UsernameField.text!, DataKeys.Password: PasswordField.text!]
        
        // store the data in the keychain
		keychain.StoreKeyValues(trackerData)
		
		let dataValid = TrackerService.Shared.setCurrentTrackerType(.DoneDone)
		
		if !dataValid {
			Log.error("DoneDoneLoginViewController invalid data", trackerData)
			return
		}
        
        //var tracker = TrackerService.Shared.getTrackerInstance(TrackerType.DoneDone)
		
        TrackerService.Shared.currentTracker?.authenticate(data: trackerData) { authResult in
            
            switch authResult {
            case let .Error(error):
                // failure
                if error.domain == HttpStatus.type {
					
					let message = "Unable to login to DoneDone with the information provided\n" +
						((error.code == HttpStatus.Unauthorized.rawValue || error.code == HttpStatus.PaymentRequired.rawValue) ? "\nReview Username & Password\nthen try again" : error.code == HttpStatus.NotFound.rawValue ? "\nReview Subdomain then try again" : "")
						
					let errorDialog = UIAlertController(title: "Error", message: message, preferredStyle: UIAlertControllerStyle.Alert)
						errorDialog.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.Cancel, handler: nil))
					
					self.presentViewController(errorDialog, animated: true) {
						
						let trackerData = [DataKeys.Subdomain: "", DataKeys.UserName: "", DataKeys.Password: ""]
						
						// clear the data in the keychain
						self.keychain.StoreKeyValues(trackerData)
					}
                }
            case .Value(_):
				// This isn't right
				self.presentingViewController?.dismissViewControllerAnimated(true, completion: nil)
            }
			
			self.ActivityIndicator.stopAnimating()
        }
    }
}